File: LABS/LAB_02/src/ki304/vasylkiv/lab2/Person.java
package ki304.vasylkiv.lab2;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;

/**
 * Class <code>Person</code> is a model of a person with personal information,
 *  address, health status, and activity status.
* @author Rostyslav Vasylkiv
 * @version 1.0
 */
 

public class Person {
	
    private PersonalInfo personalInfo;
    private Address address;
    private HealthStatus healthStatus;
    private ActivityStatus activityStatus;
    private PrintWriter fout;

    /**
     * Constructor
     * 
     * @throws FileNotFoundException
     */
    public Person() throws FileNotFoundException {
    	personalInfo = new PersonalInfo();
    	address = new Address();
    	healthStatus = new HealthStatus();
    	activityStatus = new ActivityStatus();

    	fout = new PrintWriter(new File("Log.txt"));
    }

    /**
     * Constructor
     * 
     * @param newFirstName First name
     * @param newLastName  Last name
     * @param age           Age
     * @param newCountry    Country of residence
     * @param newCity       City of residence
     * @param newStreet     Street of residence
     * @param newHouseNumber House number of residence
     * @param newWeight     Weight
     * @param newHeight     Height
     * @throws FileNotFoundException 
     */
    public Person(String newFirstName, String newLastName, int age, String newCountry, String newCity,
    		String newStreet, int newHouseNumber, double newWeight, double newHeight) throws FileNotFoundException {
    	personalInfo = new PersonalInfo(newFirstName, newLastName, age);
    	address = new Address(newCountry, newCity, newStreet, newHouseNumber);
    	healthStatus = new HealthStatus(newWeight, newHeight);
    	activityStatus = new ActivityStatus();

    	fout = new PrintWriter(new File("Log.txt"));
    }
    
    /**
     * Method changes personal information and logs the update.
     *
     * @param newFirstName New first name
     * @param newLastName  New last name
     * @param newAge       New age
     */ 
   public void changePersonalInfo(String newFirstName, String newLastName, int newAge) {
  	  personalInfo.setFirstName(newFirstName);
  	  personalInfo.setLastName(newLastName);
  	  personalInfo.setAge(newAge);
  	  fout.println("Personal data has been changed");
      fout.println();
      fout.flush();
    }
   
   /**
    * Method changes address information and logs the update.
    *
    * @param newCountry    New country of residence
    * @param newCity       New city of residence
    * @param newStreet     New street of residence
    * @param newHouseNumber New house number of residence
    */
   public void changeAddress(String newCountry, String newCity, String newStreet, int newHouseNumber) {
	   address.setCountry(newCountry);
	   address.setCity(newCity);
	   address.setStreet(newStreet);
	   address.setHouseNumber(newHouseNumber);
	   fout.println("Address of residence changed");
	   fout.println();
	   fout.flush();
   } 
   
   /**
    * Method changes weight and height information, updates BMI, and logs the update.
    *
    * @param newWeigth New weight
    * @param newHeight New height
    */
  public void changeWeigthAndHeight(double newWeigth, double newHeight) {
	  healthStatus.setWeight(newWeigth);
	  healthStatus.setHeight(newHeight);
	  healthStatus.updateBMI();
	  fout.println("Weight and height data changed");
      fout.println();
      fout.flush();
  }
  
  /**
   * Method changes activity status based on the provided status code and logs the update.
   *
   * @param status Status code representing the new activity status
   */
  public void changeActivityStatus(int status) {
      if (status == 1) {
    	  activityStatus.setMovingActivity();
      } else if (status == 2) {
    	  activityStatus.setStangingActivity();
      }
        else if (status == 3) {
          activityStatus.setSittingActivity();
      } else {
    	  activityStatus.setLieActivity();
      }

	  fout.println("Activity mode changed to " + activityStatus.getCurrentActivity());
      fout.println();
      fout.flush();
  }
  
  /**
   * Method simulates the person growing up and logs the update.
   */
  public void growUp() {
      personalInfo.growUp();
      fout.println(personalInfo.getFullName() + "has grown up.");
      fout.println();
      fout.flush();
  }
  
  /**
   * Method provides personal info in logs
   */
  public void getPersonalInfo() {
      fout.println("First name: " + personalInfo.getFirstName());
      fout.println("Last name: " + personalInfo.getLastName());
      fout.println("Age: " + personalInfo.getAge());
      fout.println();
      fout.flush();
  }
  
  /**
   * Method provides address info in logs
   */
  public void getAddresInfo() {
      fout.println(personalInfo.getFullName() + "'s Address: " + address.getToString());
      fout.println();
      fout.flush();
  }
 
  /**
   * Method provides health status in logs
   */
  public void getHealthStatus() {
	  fout.println(healthStatus.checkHealthStatus());
      fout.println();
      fout.flush();
  }
  
  /**
   * Method provides information about weight and height in logs
   */
  public void getWeightAndHeight() {
	  fout.println("Body weight: " + healthStatus.getWeight());
	  fout.println("Height: " + healthStatus.getHeight());
      fout.println();
      fout.flush();
  }
  
  /**
   * Method provides current activity in logs
   */
  public void getCurrentActivity() {
	  fout.println("The activity mode is set to " + activityStatus.getCurrentActivity());
      fout.println();
      fout.flush(); 
  }
  
  /**
   * Method provides all info about the person in logs
   */
  public void getAllPersonInfo() {
  	  fout.println("ALL INFORMATION ABOUT THE PERSON: ");
  	  fout.println();
  	  fout.println("First name: " + personalInfo.getFirstName());
      fout.println("Last name: " + personalInfo.getLastName());
      fout.println("Age: " + personalInfo.getAge());
      fout.println("Address: " + address.getToString());
      fout.println(healthStatus.checkHealthStatus());
      fout.println();
  	  fout.flush(); 
    }
   
  
  /**
   * Method releases used recourses
   */
  public void cleanUp() {
      fout.close();
  }
}

File: LABS/LAB_09/Person.py
from Address import Address
from PersonalInfo import PersonalInfo
from HealthStatus import HealthStatus
from ActivityStatus import ActivityStatus

class Person:
    
    # Constructor
    def __init__(self, firstName="unknown", lastName="unknown", age=0, country="unknown", city="unknown", street="unknown", houseNumber=0, weight=1.0, height=1.0):
        self.__personalInfo = PersonalInfo(firstName, lastName, age)
        self.__address = Address(country, city, street, houseNumber)
        self.__healthStatus = HealthStatus(weight, height)
        self.__activityStatus = ActivityStatus()

    # Method: Changes the personal information of the Person object.
    def changePersonalInfo(self, firstName, lastName, age):
        self.__personalInfo.setFirstName(firstName)
        self.__personalInfo.setLastName(lastName)
        self.__personalInfo.setAge(age)
        print("Personal data has been changed")
        print()

    # Method: Changes the address of residence of the Person object.
    def changeAddress(self, newCountry, newCity, newStreet, newHouseNumber):
        self.__address.setCountry(newCountry)
        self.__address.setCity(newCity)
        self.__address.setStreet(newStreet)
        self.__address.setHouseNumber(newHouseNumber)
        print("Address of residence changed")
        print()

    # Method: Changes the weight and height data of the Person object.
    def changeWeightAndHeight(self, newWeight, newHeight):
        self.__healthStatus.setWeight(newWeight)
        self.__healthStatus.setHeight(newHeight)
        self.__healthStatus.updateBMI()
        print("Weight and height data changed")
        print()

    # Method: Changes the activity mode of the Person object based on the provided status.
    def changeActivityStatus(self, status):
        if status == 1:
            self.__activityStatus.setMovingActivity()
        elif status == 2:
            self.__activityStatus.setStandingActivity()
        elif status == 3:
            self.__activityStatus.setSittingActivity()
        else:
            self.__activityStatus.setLieActivity()

        print("Activity mode changed to " + str(self.__activityStatus.getCurrentActivity()))
        print()

    # Method: Simulates the Person object growing up and prints a message.
    def growUp(self):
        self.__personalInfo.growUp()
        print(str(self.__personalInfo.getFullName()) + " has grown up.")
        print()

    # Method: Prints the personal information of the Person object.
    def getPersonalInfo(self):
        print("First name: " + str(self.__personalInfo.getFirstName()))
        print("Last name: " + str(self.__personalInfo.getLastName()))
        print("Age: " + str(self.__personalInfo.getAge()))
        print()

    # Method: Prints the address information of the Person object.
    def getAddressInfo(self):
        print(str(self.__personalInfo.getFullName()) + "'s Address: " + str(self.__address.toString()))
        print()

    # Method: Prints the health status of the Person object.
    def getHealthStatus(self):
        print(self.__healthStatus.checkHealthStatus())
        print()

    # Method: Prints the weight and height information of the Person object.
    def getWeightAndHeight(self):
        print("Body weight: " + str(self.__healthStatus.getWeight()))
        print("Height: " + str(self.__healthStatus.getHeight()))
        print()

    # Method: Prints the current activity mode of the Person object.
    def getCurrentActivity(self):
        print("The activity mode is set to " + str(self.__activityStatus.getCurrentActivity()))
        print()

    # Method: Prints all information about the Person object, including personal information, address, health status, and current activity.
    def getAllPersonInfo(self):
        print("ALL INFORMATION ABOUT THE PERSON:\n")
        print("First name: " + str(self.__personalInfo.getFirstName()))
        print("Last name: " + str(self.__personalInfo.getLastName()))
        print("Age: " + str(self.__personalInfo.getAge()))
        print("Address: " + str(self.__address.toString()))
        print(self.__healthStatus.checkHealthStatus())
        print("Live: " + str(self.__activityStatus.getCurrentActivity()))
        print()
